src/tileiterator.jl

```diff
@@ -0,0 +1,162 @@
+################################################################################
+##### CoveredRange
+################################################################################
+
+# a range covered by subranges
+struct CoveredRange{R,S} <: AbstractVector{UnitRange{Int}}
+    offsets::R
+    stopping::S
+end
+
+struct FixedLength
+    length::Int
+end
+
+struct LengthAtMost{N <: Union{Int, Nothing}}
+    maxlength::Int
+    maxstop::N
+end
+
+function compute_stop(offset, stopping::FixedLength)
+    return offset+stopping.length
+end
+function compute_stop(offset, stopping::LengthAtMost)
+    return min(offset+stopping.maxlength, stopping.maxstop)
+end
+
+compute_range(offset, stopping)::UnitRange{Int} = (offset+1):compute_stop(offset, stopping)
+
+Base.size(o::CoveredRange,args...) = size(o.offsets, args...)
+Base.@propagate_inbounds function Base.getindex(o::CoveredRange, inds...)
+    offset = o.offsets[inds...]
+    return compute_range(offset, o.stopping)
+end
+
+################################################################################
+##### RoundedRange
+################################################################################
+struct RoundedRange{R} <: AbstractVector{Int}
+    range::R
+end
+
+Base.@propagate_inbounds function Base.getindex(r::RoundedRange, i)
+    rough = r.range[i]
+    return round(Int, rough)
+end
+
+function roundedrange(start; stop, length)
+    inner = LinRange(start, stop, length)
+    return RoundedRange(inner)
+end
+
+Base.size(o::RoundedRange, args...) = size(o.range, args...)
+
+################################################################################
+##### TileIterator
+################################################################################
+struct TileIterator{N,C} <: AbstractArray{NTuple{N, UnitRange{Int}}, N}
+    covers1d::C
+end
+
+function TileIterator(covers1d::NTuple{N, AbstractVector{UnitRange{Int}}}) where {N}
+    C = typeof(covers1d)
+    return TileIterator{N, C}(covers1d)
+end
+
+function TileIterator(axes::Indices{N}, tilesize::Dims{N}) where {N}
+    tileiterator(axes, RelaxLastTile(tilesize))::TileIterator{N}
+end
+
+# strategies
+export RelaxStride
+struct RelaxStride{N}
```

Comment:
I thought, lets be conservative for now. I just added these two strategies. They are enough to cover the funcionality on `master` as well as my balanced tile size use case.

Reply:
A different PR could implement a strategy that sets the stride to `(1,1,1...)` giving the behavior sketched here:
https://github.com/JuliaImages/ImageFiltering.jl/pull/155#issuecomment-706649346

Reply:
For more advanced strides, we might want to do
`tileiterator(axes, UnitStride())[begin:2:end, :, begin+2:4:end-2]`

Reply:
This works but might not be the most efficient way because indexing with `[begin:2:end, :, begin+2:4:end-2]` allocates a new array. No need to implement in this PR, though.

Reply:
Yes you are right this allocates currently, we can fix it, when it becomes a practical obstacle.